File: packages/kotlin/src/test/kotlin/com/openbancor/catalogmx/generators/IdentityGeneratorTest.kt
package com.openbancor.catalogmx.generators

import com.openbancor.catalogmx.validateRFC
import com.openbancor.catalogmx.validateCURP
import com.openbancor.catalogmx.validateCLABE
import com.openbancor.catalogmx.validateNSS
import com.openbancor.catalogmx.generatePersonaFisica
import com.openbancor.catalogmx.generatePersonaMoral
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import kotlin.random.Random

class IdentityGeneratorTest {

    @Test
    fun `should generate valid persona fisica`() {
        val generator = IdentityGenerator(Random(42)) // Fixed seed for reproducibility
        val persona = generator.generatePersonaFisica()

        // Verify basic fields are present
        assertTrue(persona.nombre.isNotEmpty())
        assertTrue(persona.apellidoPaterno.isNotEmpty())
        assertTrue(persona.apellidoMaterno.isNotEmpty())
        assertTrue(persona.nombreCompleto.isNotEmpty())

        // Verify RFC is valid
        assertEquals(13, persona.rfc.length)
        assertTrue(validateRFC(persona.rfc), "RFC should be valid: ${persona.rfc}")

        // Verify CURP is valid
        assertEquals(18, persona.curp.length)
        assertTrue(validateCURP(persona.curp), "CURP should be valid: ${persona.curp}")

        // Verify NSS is valid
        assertEquals(11, persona.nss.length)
        assertTrue(validateNSS(persona.nss), "NSS should be valid: ${persona.nss}")

        // Verify CLABE is valid
        assertEquals(18, persona.clabe.length)
        assertTrue(validateCLABE(persona.clabe), "CLABE should be valid: ${persona.clabe}")

        // Verify age is within expected range
        assertTrue(persona.edad in 18..65)

        // Verify sexo is valid
        assertTrue(persona.sexo in listOf("H", "M"))
    }

    @Test
    fun `should generate persona fisica with specified sex`() {
        val generator = IdentityGenerator(Random(42))

        val male = generator.generatePersonaFisica(sexo = "H")
        assertEquals("H", male.sexo)
        assertEquals("Masculino", male.sexoDescripcion)

        val female = generator.generatePersonaFisica(sexo = "M")
        assertEquals("M", female.sexo)
        assertEquals("Femenino", female.sexoDescripcion)
    }

    @Test
    fun `should generate persona fisica with specified state`() {
        val generator = IdentityGenerator(Random(42))

        val persona = generator.generatePersonaFisica(estado = "JC")
        assertEquals("JC", persona.estadoNacimientoCode)
        assertEquals("JALISCO", persona.estadoNacimiento)
    }

    @Test
    fun `should generate persona fisica with custom age range`() {
        val generator = IdentityGenerator(Random(42))

        val persona = generator.generatePersonaFisica(minAge = 25, maxAge = 30)
        assertTrue(persona.edad in 25..30, "Age ${persona.edad} should be between 25 and 30")
    }

    @Test
    fun `should generate valid persona moral`() {
        val generator = IdentityGenerator(Random(42))
        val persona = generator.generatePersonaMoral()

        // Verify basic fields are present
        assertTrue(persona.razonSocial.isNotEmpty())
        assertTrue(persona.nombreComercial.isNotEmpty())

        // Verify RFC is valid (12 chars for moral)
        assertEquals(12, persona.rfc.length)
        assertTrue(validateRFC(persona.rfc), "RFC should be valid: ${persona.rfc}")

        // Verify CLABE is valid
        assertEquals(18, persona.clabe.length)
        assertTrue(validateCLABE(persona.clabe), "CLABE should be valid: ${persona.clabe}")

        // Verify antiguedad is within expected range
        assertTrue(persona.antiguedad in 1..30)
    }

    @Test
    fun `should generate persona moral with custom age range`() {
        val generator = IdentityGenerator(Random(42))

        val persona = generator.generatePersonaMoral(minYears = 5, maxYears = 10)
        assertTrue(persona.antiguedad in 5..10, "Antiguedad ${persona.antiguedad} should be between 5 and 10")
    }

    @Test
    fun `convenience functions should work`() {
        val fisica = generatePersonaFisica()
        assertNotNull(fisica)
        assertTrue(fisica.containsKey("rfc"))
        assertTrue(fisica.containsKey("curp"))
        assertTrue(fisica.containsKey("nss"))
        assertTrue(fisica.containsKey("clabe"))

        val moral = generatePersonaMoral()
        assertNotNull(moral)
        assertTrue(moral.containsKey("rfc"))
        assertTrue(moral.containsKey("clabe"))
    }

    @Test
    fun `should include complete address`() {
        val generator = IdentityGenerator(Random(42))
        val persona = generator.generatePersonaFisica()

        val direccion = persona.direccion
        assertTrue(direccion.containsKey("calle"))
        assertTrue(direccion.containsKey("numero_exterior"))
        assertTrue(direccion.containsKey("colonia"))
        assertTrue(direccion.containsKey("codigo_postal"))
        assertTrue(direccion.containsKey("municipio"))
        assertTrue(direccion.containsKey("estado"))
    }

    @Test
    fun `should include bank information`() {
        val generator = IdentityGenerator(Random(42))
        val persona = generator.generatePersonaFisica()

        val banco = persona.banco
        assertTrue(banco.containsKey("code"))
        assertTrue(banco.containsKey("name"))
        assertTrue(banco["code"]!!.length == 3)
    }

    @Test
    fun `should include fiscal regime`() {
        val generator = IdentityGenerator(Random(42))
        val persona = generator.generatePersonaFisica()

        val regimen = persona.regimenFiscal
        assertTrue(regimen.containsKey("code"))
        assertTrue(regimen.containsKey("name"))
    }

    @Test
    fun `should generate valid email format`() {
        val generator = IdentityGenerator(Random(42))
        val persona = generator.generatePersonaFisica()

        assertTrue(persona.email.contains("@"))
        assertTrue(persona.email.contains("."))
    }

    @Test
    fun `should generate valid phone number`() {
        val generator = IdentityGenerator(Random(42))
        val persona = generator.generatePersonaFisica()

        assertEquals(10, persona.telefono.length)
        assertTrue(persona.telefono.all { it.isDigit() })
    }

    @Test
    fun `toMap should produce valid JSON-serializable structure`() {
        val generator = IdentityGenerator(Random(42))
        val persona = generator.generatePersonaFisica()

        val map = persona.toMap()

        assertEquals("persona_fisica", map["tipo"])
        assertEquals(persona.rfc, map["rfc"])
        assertEquals(persona.curp, map["curp"])
        assertEquals(persona.nss, map["nss"])
        assertEquals(persona.clabe, map["clabe"])
    }
}

File: packages/dart/test/identity_generator_test.dart
import 'package:catalogmx/catalogmx.dart';
import 'package:test/test.dart';

void main() {
  group('IdentityGenerator', () {
    test('generatePersonaFisica returns valid identity', () {
      final identity = generatePersonaFisica();

      expect(identity['tipo'], equals('persona_fisica'));
      expect(identity['nombre'], isNotEmpty);
      expect(identity['apellido_paterno'], isNotEmpty);
      expect(identity['apellido_materno'], isNotEmpty);
      expect(identity['rfc'], isNotEmpty);
      expect(identity['curp'], isNotEmpty);
      expect(identity['nss'], isNotEmpty);
      expect(identity['clabe'], isNotEmpty);

      // Validate RFC
      expect(validateRFC(identity['rfc'] as String), isTrue);

      // Validate CURP
      expect(validateCURP(identity['curp'] as String), isTrue);

      // Validate NSS
      expect(validateNSS(identity['nss'] as String), isTrue);

      // Validate CLABE
      expect(validateCLABE(identity['clabe'] as String), isTrue);
    });

    test('generatePersonaFisica with male sex', () {
      final identity = generatePersonaFisica(sexo: 'H');

      expect(identity['sexo'], equals('H'));
      expect(identity['sexo_descripcion'], equals('Masculino'));

      // CURP should have H for male
      final curp = identity['curp'] as String;
      expect(curp[10], equals('H'));
    });

    test('generatePersonaFisica with female sex', () {
      final identity = generatePersonaFisica(sexo: 'M');

      expect(identity['sexo'], equals('M'));
      expect(identity['sexo_descripcion'], equals('Femenino'));

      // CURP should have M for female
      final curp = identity['curp'] as String;
      expect(curp[10], equals('M'));
    });

    test('generatePersonaFisica with specific state', () {
      final identity = generatePersonaFisica(estado: 'JC');

      expect(identity['estado_nacimiento_code'], equals('JC'));
      expect(
          (identity['estado_nacimiento'] as String)
              .toUpperCase()
              .contains('JALISCO'),
          isTrue);
    });

    test('generatePersonaMoral returns valid identity', () {
      final identity = generatePersonaMoral();

      expect(identity['tipo'], equals('persona_moral'));
      expect(identity['razon_social'], isNotEmpty);
      expect(identity['nombre_comercial'], isNotEmpty);
      expect(identity['rfc'], isNotEmpty);
      expect(identity['clabe'], isNotEmpty);

      // Validate RFC (should be 12 characters for moral)
      final rfc = identity['rfc'] as String;
      expect(rfc.length, equals(12));
      expect(validateRFC(rfc), isTrue);

      // Validate CLABE
      expect(validateCLABE(identity['clabe'] as String), isTrue);
    });

    test('generateIdentity with tipo fisica', () {
      final identity = generateIdentity(tipo: 'fisica');
      expect(identity['tipo'], equals('persona_fisica'));
      expect(identity['curp'], isNotEmpty);
    });

    test('generateIdentity with tipo moral', () {
      final identity = generateIdentity(tipo: 'moral');
      expect(identity['tipo'], equals('persona_moral'));
      expect(identity['razon_social'], isNotEmpty);
    });

    test('IdentityGenerator class generates valid data', () {
      final generator = IdentityGenerator();

      final fisica = generator.generatePersonaFisica();
      expect(fisica.tipo, equals('persona_fisica'));
      expect(validateRFC(fisica.rfc), isTrue);
      expect(validateCURP(fisica.curp), isTrue);
      expect(validateNSS(fisica.nss), isTrue);
      expect(validateCLABE(fisica.clabe), isTrue);

      final moral = generator.generatePersonaMoral();
      expect(moral.tipo, equals('persona_moral'));
      expect(validateRFC(moral.rfc), isTrue);
      expect(validateCLABE(moral.clabe), isTrue);
    });

    test('multiple generations produce unique identities', () {
      final identities = <String>{};

      for (var i = 0; i < 10; i++) {
        final identity = generatePersonaFisica();
        final rfc = identity['rfc'] as String;
        identities.add(rfc);
      }

      // Most should be unique (very unlikely to have duplicates)
      expect(identities.length, greaterThan(5));
    });

    test('age constraints are respected', () {
      final identity = generatePersonaFisica(minAge: 25, maxAge: 30);
      final age = identity['edad'] as int;

      expect(age, greaterThanOrEqualTo(25));
      expect(age, lessThanOrEqualTo(30));
    });

    test('PersonaFisica toJson returns all fields', () {
      final generator = IdentityGenerator();
      final fisica = generator.generatePersonaFisica();
      final json = fisica.toJson();

      expect(json.containsKey('tipo'), isTrue);
      expect(json.containsKey('nombre'), isTrue);
      expect(json.containsKey('apellido_paterno'), isTrue);
      expect(json.containsKey('apellido_materno'), isTrue);
      expect(json.containsKey('nombre_completo'), isTrue);
      expect(json.containsKey('fecha_nacimiento'), isTrue);
      expect(json.containsKey('edad'), isTrue);
      expect(json.containsKey('sexo'), isTrue);
      expect(json.containsKey('rfc'), isTrue);
      expect(json.containsKey('curp'), isTrue);
      expect(json.containsKey('nss'), isTrue);
      expect(json.containsKey('banco'), isTrue);
      expect(json.containsKey('clabe'), isTrue);
      expect(json.containsKey('direccion'), isTrue);
      expect(json.containsKey('telefono'), isTrue);
      expect(json.containsKey('email'), isTrue);
      expect(json.containsKey('regimen_fiscal'), isTrue);
    });

    test('PersonaMoral toJson returns all fields', () {
      final generator = IdentityGenerator();
      final moral = generator.generatePersonaMoral();
      final json = moral.toJson();

      expect(json.containsKey('tipo'), isTrue);
      expect(json.containsKey('razon_social'), isTrue);
      expect(json.containsKey('nombre_comercial'), isTrue);
      expect(json.containsKey('fecha_constitucion'), isTrue);
      expect(json.containsKey('antiguedad'), isTrue);
      expect(json.containsKey('rfc'), isTrue);
      expect(json.containsKey('banco'), isTrue);
      expect(json.containsKey('clabe'), isTrue);
      expect(json.containsKey('direccion'), isTrue);
      expect(json.containsKey('telefono'), isTrue);
      expect(json.containsKey('email'), isTrue);
      expect(json.containsKey('regimen_fiscal'), isTrue);
    });
  });
}
